### ISSUE TITLE
Dynamically mapping DTO from Rest Webservice (@JsonFilter)

### ISSUE THREAD

**Issue by axelpcr** [mapstruct/mapstruct#788]
Hi guys,

I Would like to know if is possible to mapping dynamically DTO.

Example : 

My Webservice
``` Java

        @GET
	@Produces(MediaType.APPLICATION_JSON)
	@Path("clients")
	public String getClients() throws JsonProcessingException {

		ObjectMapper mapper = new ObjectMapper();
		FilterProvider filters = new SimpleFilterProvider().addFilter("myFilter",
				SimpleBeanPropertyFilter.filterOutAllExcept("name", "surname", "city"));
		mapper.setFilters(filters);
		List<ClientDto> clients = clientService.mapClientDtos(clientService.lister());

		return mapper.writer(filters).writeValueAsString(clients);
	}
```

My DTO

``` Java
@JsonFilter("myFilter")
public class ClientDto {

	private Integer idClient;
	private String name;
	private String surname;
	private String city;
	private String telephone;
	private String portable;

Getters & Setters
// //
```

My Mapper
``` Java

        @Mapping(source = "idClient", target = "idClient"),
	@Mapping(source = "name", target = "name"),
	@Mapping(source = "surname", target = "surname"),
	@Mapping(source = "city", target = "city"),
	@Mapping(source = "telephone", target = "telephone"),
	@Mapping(source = "portable", target = "portable")})
	public abstract ClientDto clientToClientDto(Client client);
```

It's possible of any way to map just the properties I need from my Webservice (@JsonFilter) ?

I seeing Orika can be that :

> Determine if a given field mapping should be performed
>  http://orika-mapper.github.io/orika-docs/filters.html
> The shouldMap(Type<?> sourceType, String sourceName, Type<?> destType, String destName, MappingContext mappingContext) method is called to determine whether the mapping should be performed on an individual case-by-case basis. This allows the filter to make a dynamic decision at runtime as to whether the fields should be mapped.

But I prefer MapStruct and I would like stay on this technology.

Thx :)

**Comment by gunnarmorling** [comment]
Hey @axelpcr, that's not something explicitly supported by MapStruct atm.

I don't know about `@JsonFilter`, is the decision whether to map a given property or not one you can make at compile time (i.e. the time when MapStruct generates code) or is it only possible to decide at runtime? I can see us adding an SPI for the former; The latter would be harder to achieve. Maybe you could use an `@AfterMapping` method to unset specific properties?

Out of interest, why do you have properties in your DTO if you don't want to populate them (IIUC)? You don't need the `@Mapping` annotations btw. in case source and target property are the same.

**Comment by heitmanm** [comment]
I think @axelpcr is searching for something like entity-filtering.

https://github.com/jersey/jersey/tree/master/examples/entity-filtering
https://github.com/FasterXML/jackson-databind/issues/1040

I also tried to implement it in my JHipster / SpringBoot application without any success at the moment so I had the idea to use mapstruct for this filtering. 

@axelpcr do you already have an implementation for the feature?

**Comment by agudian** [comment]
Would be interesting to know what the generated code is supposed to look like. Perhaps that can bring the discussion a bit forward... Right now I still don't have an idea of how this feature would be used exactly... :-)

**Comment by axelpcr** [comment]
@heitmanm : Yeah, I use Jersey entity-filtering now and it's work good.
You say you have a problem with entity-filetring, try this tutorial :
http://www.codingpedia.org/ama/restful-web-services-example-in-java-with-jersey-spring-and-mybatis/

Generally the problems with entity-filtering are jersey andjackson dependencies. I upgrade my Java 1.6 to Java 1.7 and Jackson1 to Jackson2 and now it's work.

I try to explain better what I want.

Imagine, I have one generic Dto Object, ClientDto which map Client entity.

``` Java
public class ClientDto {

	private Integer idClient;
	private String raisonSociale;
	private String responsable;
	private String libelleInterne;
	private String telephone;

//Getters & Setters
```

My Jersey Webservice

``` Java
	@GET
	@Produces(MediaType.APPLICATION_JSON)
	@Path("clients")
	public Response getClients() {

		List<ClientDto> clientDtos = new ArrayList<ClientDto>();

		for (Client client : clientService.lister()) {
			clientDtos.add(clientMapper.clientToClientDtoForCrm(client));
		}
		GenericEntity<List<ClientDto>> entity = new GenericEntity<List<ClientDto>>(clientDtos) {
		};
		return Response.ok(entity).build();
	}
```
With Jersey entity-filtering, the http client can choose which attributs he would like.

If http client enter this url :

> localhost:8080/webservices/clients

My Webservice return that:

> {
>    "idClient": "1",
>    "raisonSociale": "Gnu",
>    "responsable": "Stallman",
>    "libelleInterne": "Gnu",
>    "telephone": "06.66.66.66.66"
> }

If http client enter this url :

> localhost:8080/webservices/clients?select=idClient,raisonSociale

My Webservice  return that:

> {
>    "idClient": "1",
>    "raisonSociale": "Gnu"
> }

Interest is to have a dynamic generic webservice.That's why I want just to make 1 generic Dto.
I don't know which attributs http client needed.

My first problem is resolved by Jersey entity-filtering, http client gets only needed attributs of the Dto.

But what's  the interest to map attributes which http client don't need.

I programm a big data synchronisation and my webservice provide a hudge Clients list.

I would like to do a dynamic mapping at runtime, for increase performance of my programm.

I hope i'm explicit.

Thx :)

**Comment by agudian** [comment]
OK, I get the use case and how you'd like to use the results. What I'm missing is how you want to pass the filter-information to the mapper and how the generated mapper implementation is supposed to use that information.

All in all, I could imagine that this is kind-of the same problem space as described in #778, where some other means are to be used in the mapper implementation to decide if a certain property is mapped or not. So what I think we need is an SPI to more easily add customizations to the generated code, such as wrapping the individual property mappings in the method with a check in your JsonFilter settings or in the Hibernate helper method. That way we don't need to pull in every kind of technology into MapStruct itself, but can provide extensions for edge-cases.

Would that work for you? It kind of goes alongside your Jackson example, where you'd also add an additional module to tie Jersey and Jackson together (if I understood that correctly).

**Comment by gunnarmorling** [comment]
@axelpcr, when you say "generic DTO", it is still not a map-kind structure but a strongly typed DTO, it's just that you want to dynamically select which properties to populate, right?

**Comment by axelpcr** [comment]
Yeah, I just want to populate specific properties. This properties needs to chosen at runtime.

**Comment by jmuis** [comment]
Hi, 

Are there currently any concrete plans to implement a SPI to filter in/out properties?
We have just taken mapstruct on board, but this is kind of a showstopper for us, as we really need to filter out 'recursive' structures based on real-time user input as we typically don't need full recursion on the mapping, but sometimes do. It all depends on context...

I would be honoured to help implement / design it but as a new user to mapstruct the learning curve is pretty steep, and I really wouldn't like to fork and maintain my own version of mapstruct having a custom feature that would work just for me.

In general my idea would be to implement an annotation on the target property and a '3rd' mapping parameter providing the SPI / mapping context where, similar to the lifecycle callbacks that have been implemented already, just on a property based level.

Thanks,
Jeroen

**Comment by joinghel** [comment]
I have a similar use-case which would be really cool. Assume an admin is allowed to update the license of a `Car`, but a normal user is not. I would add a `ReadOnly` annotation on the license property indicating that the field is readOnly for a non-admin user:

```Java
CarDTO {
    Integer numberOfWheels;
    @ReadOnly(role = Role.USER)
    String license;
}
```

The mapping method would be pretty easy:

```Java
Car toCar(CarDTO, @Context Role role);
```

Some generic function could then be defined to evaluate if an attribute should be mapped:

```Java
default boolean shouldMap(javax.validation.metadata.PropertyDescriptor sourceProperty, @Context Role role) {
    // Loop over property annotations
    // Check if property has annotation ReadOnly && annotation.role != role
    return shouldMap;
}
```

Right now, the only solution is to define two mappers: one for the normal user with `ignore=true` for license and one for the admin without any ignores.

**Comment by Baravd** [comment]
> Would also like to have a similar feature. The reason for me is performance and optimization. I use a combination of JPA and Mapstruct.
> 
> Consider this example. My JPA entity has multiple child entities that are lazy-loaded when the getter is called on an entity. Each call to getter of a lazily-loaded entity results in a separate DB call. (One option is to load them eargerly but that creates a problem of pulling lots of data down the chain while it may not be needed).
> 
> Mapstruct by default calls all the getters on the entity to set the values on the DTO object. But client might not always need every attribute value of the DTO. I want give the option on my client side to pass a list of attributes that they need. I'll pass this as afilter to mapstruct and it can simply omit calling the getter on the DTO which is not request by the filter.

Totally agree. The workaround I did is to create multiple DTOs and multiple mappers  for each scenario. For example on some endpoint I want to return just basic attributes from a Product entity(id, price, name, description). So I created BasciProductDto and a mapper. In the second situation I want also to send all the information (for example reviews which is a OneToMany relationship). So I created  FullProductDto and a mapper.

Obviously this can get messy having to many mappers, but I just extracted them to a common class and used that in the Controller.

Hope this helps.

**Comment by Israelknowing** [comment]
> [@axelpcr](https://github.com/axelpcr), when you say "generic DTO", it is still not a map-kind structure but a strongly typed DTO, it's just that you want to dynamically select which properties to populate, right?





https://stackdev.space/questions/optimal-approach-for-managing-child-information-array-within-a-rest-api
